scipy/stats/stats.py

```diff
@@ -5117,24 +5346,105 @@ def ks_2samp(data1, data2):
-    try:
-        prob = distributions.kstwobign.sf((en + 0.12 + 0.11 / en) * d)
-    except Exception:
-        warnings.warn('This should not happen! Please open an issue at '
-                    'https://github.com/scipy/scipy/issues and provide the code '
-                    'you used to trigger this warning.\n')
-        prob = 1.0
+    cddiffs = cdf1 - cdf2
+    minS = -np.min(cddiffs)
+    maxS = np.max(cddiffs)
+    alt2Dvalue = {'less': minS, 'greater': maxS, 'two-sided': max(minS, maxS)}
+    d = alt2Dvalue[alternative]
+    g = gcd(n1, n2)
+    n1g = n1 // g
+    n2g = n2 // g
+    prob = -np.inf
+    original_mode = mode
+    if mode == 'auto':
+        if max(n1, n2) <= LARGE_N:
+            mode = 'exact'
+        else:
+            mode = 'asymp'
+    elif mode == 'exact':
+        # If lcm(n1, n2) is too big, switch from exact to asymp
+        if n1g >= np.iinfo(np.int).max / n2g:
+            mode = 'asymp'
+            warnings.warn(
+                "Exact ks_2samp calculation not possible with samples sizes "
+                "%d and %d. Switching to 'asymp' " % (n1, n2), RuntimeWarning)
+
+    saw_fp_error = False
+    if mode == 'exact':
+        lcm = (n1 // g) * n2
+        h = int(np.round(d * lcm))
+        d = h * 1.0 / lcm
+        if h == 0:
+            prob = 1.0
+        else:
+            try:
+                if alternative == 'two-sided':
+                    if n1 == n2:
+                        prob = _compute_prob_outside_square(n1, h)
+                    else:
+                        prob = 1 - _compute_prob_inside_method(n1, n2, g, h)
+                else:
+                    if n1 == n2:
+                        # prob = binom(2n, n-h) / binom(2n, n)
+                        # Evaluating in that form incurs roundoff errors
+                        # from special.binom. Instead calculate directly
+                        prob = 1.0
+                        for j in range(h):
+                            prob = (n1 - j) * prob / (n1 + j + 1)
+                    else:
+                        num_paths = _count_paths_outside_method(n1, n2, g, h)
+                        bin = special.binom(n1 + n2, n1)
+                        if not np.isfinite(bin) or not np.isfinite(num_paths) or num_paths > bin:
+                            raise FloatingPointError()
+                        prob = num_paths / bin
+
+            except FloatingPointError:
+                # Switch mode
+                mode = 'asymp'
+                saw_fp_error = True
+                # Can't raise warning here, inside the try
+            finally:
+                if saw_fp_error:
+                    if original_mode == 'exact':
+                        warnings.warn(
+                            "ks_2samp: Exact calculation overflowed. "
+                            "Switching to mode=%s" % mode, RuntimeWarning)
+                else:
+                    if prob > 1 or prob < 0:
+                        mode = 'asymp'
```

Comment:
control flow never reaches this branch in tests

Reply:
It is protection against accumulation of floating point rounding errors taking the result out of the interval [0, 1], especially as `special.binom(a, b)` is not integral. But an example would be good.